File: crates/nil-android/android/NilVpnPlugin.kt
package com.nilvpn

import android.app.Activity
import android.content.Intent
import android.net.VpnService
import android.os.SystemClock
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import java.io.File
import java.io.RandomAccessFile
import java.nio.channels.OverlappingFileLockException

/**
 * Private Rust-to-native bridge for the `:vpn` service. Rust retains the Tauri PluginHandle and the
 * WebView has no plugin ACL, so bearer start arguments never cross JavaScript. The app process
 * redeems the blind-signed token and passes only the resulting node policy + grant here.
 */
@TauriPlugin
class NilVpnPlugin(private val activity: Activity) : Plugin(activity) {

    @Command
    fun startVpn(invoke: Invoke) {
        // OS consent: returns an Intent the first time; null once granted.
        VpnService.prepare(activity)?.let {
            activity.startActivityForResult(it, VPN_CONSENT)
            invoke.reject("VPN permission required; approve the system dialog and start again")
            return
        }
        val a = invoke.parseArgs(StartArgs::class.java)
        if (!a.reservationId.matches(Regex("[0-9a-f]{64}"))) {
            invoke.reject("invalid reservation id")
            return
        }
        val i = Intent(activity, NilVpnService::class.java).apply {
            putExtra("nodeHost", a.nodeHost)
            putExtra("nodePort", a.nodePort)
            putExtra("serverName", a.serverName ?: a.nodeHost)
            putExtra("measurementHex", a.measurementHex ?: "")
            putExtra("tlsSpkiSha256Hex", a.tlsSpkiSha256Hex ?: "")
            putExtra("transparencyLogKeyHex", a.transparencyLogKeyHex ?: "")
            putExtra("teeName", a.teeName ?: "sev-snp")
            val minTcb = a.minTcbSevsnp
            putExtra("minTcbPresent", minTcb != null)
            putExtra("minTcbFmc", minTcb?.fmc ?: -1)
            putExtra("minTcbBootloader", minTcb?.bootloader ?: 0)
            putExtra("minTcbTee", minTcb?.tee ?: 0)
            putExtra("minTcbSnp", minTcb?.snp ?: 0)
            putExtra("minTcbMicrocode", minTcb?.microcode ?: 0)
            // Coordinator grant for this hop (redeemed in the app process); "" if the path carries none.
            putExtra("grantHex", a.grantHex ?: "")
            putExtra("grantNonceHex", a.grantNonceHex ?: "")
            putExtra("reservationId", a.reservationId)
            putExtra("allowUnattested", a.allowUnattested)
        }
        // Publish "connecting" synchronously BEFORE dispatching the service, so a status poll can't
        // read a stale "down"/"up" from a previous session in the gap before the service writes.
        try {
            File(activity.filesDir, NilVpnService.STATUS_FILE)
                .writeText("${NilVpnService.CONNECTING}|${a.reservationId}|${SystemClock.elapsedRealtime()}")
        } catch (_: Throwable) { /* best-effort; the service writes it again on start */ }
        activity.startForegroundService(i)
        invoke.resolve()
    }

    /**
     * Preflight the OS VPN consent WITHOUT starting anything, so the app can obtain permission
     * BEFORE it redeems a single-use token. `{authorized:true}` means prepare() returned null
     * (already granted) — the caller may redeem + startVPN. `{authorized:false}` means consent was
     * required: we launch the system dialog and the caller must NOT redeem (it would burn a token on
     * a permission the user may still deny); the user approves, then taps Connect again.
     */
    @Command
    fun prepareVpn(invoke: Invoke) {
        val consent = VpnService.prepare(activity)
        if (consent == null) {
            invoke.resolve(JSObject().apply { put("authorized", true) })
        } else {
            activity.startActivityForResult(consent, VPN_CONSENT)
            invoke.resolve(JSObject().apply { put("authorized", false) })
        }
    }

    /**
     * Deep-link to the OS VPN settings so the user can turn on "Always-on VPN" + "Block connections
     * without VPN" — the PERSISTENT kill-switch. The app CANNOT enable this itself (PD-8 honesty);
     * all it can do is take the user there. Falls back to app details if the VPN settings screen is
     * unavailable on this device/OEM.
     */
    @Command
    fun openVpnSettings(invoke: Invoke) {
        try {
            activity.startActivity(
                Intent(android.provider.Settings.ACTION_VPN_SETTINGS)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            )
        } catch (e: Throwable) {
            activity.startActivity(
                Intent(android.provider.Settings.ACTION_SETTINGS)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            )
        }
        invoke.resolve()
    }

    @Command
    fun stopVpn(invoke: Invoke) {
        // stopService does NOT destroy a foreground VpnService (the system binds it while the TUN fd
        // is open), so deliver an explicit STOP command instead — the service tears the tunnel down
        // from inside (closing the fd unbinds the system) and stopSelf()s. startService just routes
        // the action to the already-running service's onStartCommand.
        activity.startService(Intent(activity, NilVpnService::class.java).setAction(NilVpnService.ACTION_STOP))
        invoke.resolve()
    }

    /**
     * Report the `:vpn` service's fresh tunnel state to the private Rust plugin handle:
     * `{state: connecting|up|dead|down, reservationId?: ...}`.
     * The service (a separate process) writes its health to a file in this app's shared `filesDir`;
     * we read it here. This lets the UI report "connected" only after the attestation gate actually
     * passes (state==up), and surface a dropped/stale tunnel (dead) instead of an optimistic lie.
     * Defaults to "down" when no status has been published. The random local reservation id binds
     * completion but carries no node, token, account, payment, or identity data.
     */
    @Command
    fun statusVpn(invoke: Invoke) {
        val raw = try {
            File(activity.filesDir, NilVpnService.STATUS_FILE).readText().trim()
        } catch (e: Throwable) {
            ""
        }
        val fields = raw.split('|', limit = 4)
        val recordedState = fields.getOrNull(0)?.ifEmpty { NilVpnService.DOWN }
            ?: NilVpnService.DOWN
        val reservationId = fields.getOrNull(1)?.takeIf { it.matches(Regex("[0-9a-f]{64}")) }
        val writtenAt = fields.getOrNull(2)?.toLongOrNull()
        val recordedIncarnation = fields.getOrNull(3)
            ?.takeIf { it.matches(Regex("[0-9a-f]{64}")) }
        val now = SystemClock.elapsedRealtime()
        val fresh = writtenAt != null && writtenAt <= now && now - writtenAt <= STATUS_MAX_AGE_MS
        val liveIncarnation = liveVpnServiceIncarnation()
        val sameLiveService = liveIncarnation != null && liveIncarnation == recordedIncarnation
        val recognized = recordedState in setOf(
            NilVpnService.CONNECTING,
            NilVpnService.UP,
            NilVpnService.DEAD,
            NilVpnService.DOWN,
        )
        // An old `up` record proves nothing after the service/process has died. Return `dead` so
        // Rust keeps the pending pass and tears down/retries instead of acknowledging a ghost VPN.
        val state = when {
            !recognized -> NilVpnService.DOWN
            recordedState == NilVpnService.DOWN -> NilVpnService.DOWN
            !fresh -> NilVpnService.DEAD
            // The app publishes a short connecting transition immediately before dispatching the
            // service. It never authorizes token completion, so it may precede lease acquisition.
            recordedState == NilVpnService.CONNECTING -> NilVpnService.CONNECTING
            sameLiveService -> recordedState
            else -> NilVpnService.DEAD
        }
        invoke.resolve(JSObject().apply {
            put("state", state)
            if (fresh && reservationId != null && (sameLiveService || state == NilVpnService.CONNECTING)) {
                put("reservationId", reservationId)
            }
        })
    }

    /**
     * Test the separate `:vpn` process's exclusive lease without trusting ActivityManager caches.
     * Linux releases this lock immediately on process death. The random value stored under that
     * lock binds a status record to the exact process incarnation that wrote it.
     */
    private fun liveVpnServiceIncarnation(): String? {
        var leaseFile: RandomAccessFile? = null
        return try {
            leaseFile = RandomAccessFile(File(activity.filesDir, NilVpnService.STATUS_LEASE_FILE), "rw")
            val candidate = leaseFile.channel.tryLock()
            if (candidate == null) {
                leaseFile.seek(0)
                leaseFile.readLine()?.trim()?.takeIf { it.matches(Regex("[0-9a-f]{64}")) }
            } else {
                candidate.release()
                null
            }
        } catch (_: OverlappingFileLockException) {
            try {
                leaseFile?.seek(0)
                leaseFile?.readLine()?.trim()?.takeIf { it.matches(Regex("[0-9a-f]{64}")) }
            } catch (_: Throwable) {
                null
            }
        } catch (_: Throwable) {
            null
        } finally {
            try { leaseFile?.close() } catch (_: Throwable) { /* fail-closed result already chosen */ }
        }
    }

    companion object {
        private const val VPN_CONSENT = 0x1107
        // Longer than the frontend's 20-second connection deadline, but short enough that an app
        // restart cannot mistake a long-dead service for a live tunnel.
        private const val STATUS_MAX_AGE_MS = 30_000L
    }

    class StartArgs {
        lateinit var reservationId: String
        lateinit var nodeHost: String
        var nodePort: Int = 443
        var serverName: String? = null
        var measurementHex: String? = null
        var tlsSpkiSha256Hex: String? = null
        var transparencyLogKeyHex: String? = null
        var teeName: String? = null
        var minTcbSevsnp: SevSnpTcbFloor? = null
        var grantHex: String? = null
        var grantNonceHex: String? = null
        var allowUnattested: Boolean = false
    }

    class SevSnpTcbFloor {
        var fmc: Int? = null
        var bootloader: Int = 0
        var tee: Int = 0
        var snp: Int = 0
        var microcode: Int = 0
    }
}

File: crates/nil-apple/apple/NilVpnPlugin.swift
// NilVpnPlugin.swift — the iOS Tauri v2 plugin for NIL VPN. Runs in the CONTAINER-APP process
// (not the packet-tunnel appex), and is the iOS counterpart of the Android `NilVpnPlugin.kt`.
//
// It implements five `nil-vpn` commands called through a private Rust-held PluginHandle
// (prepareVpn / startVpn / statusVpn / stopVpn / openVpnSettings); the WebView has no plugin ACL.
// The bridge drives the REAL attested
// MASQUE tunnel via `NETunnelProviderManager` + the `PacketTunnel` appex
// (`NEPacketTunnelProvider`, see `PacketTunnelProvider.swift`) — never the in-process loopback mock.
//
// Privacy (PD-3): identity NEVER reaches here. `extension_connect` (Rust, app process) already
// redeemed the blind-signed bearer token at the Coordinator; this plugin only receives the attested node
// endpoint + pinned measurement + opaque grant and forwards them to the appex via
// `providerConfiguration`. No account, email, payment, token, or destination crosses into the
// datapath. Nothing identifying is logged (PD-2).
//
// Honest kill-switch (PD-8):
//   * IN-CONNECTION fail-closed — the posture the app CAN set: the protocol's `includeAllNetworks`
//     captures all traffic into the tunnel while connected (the appex also installs the v4+v6
//     default routes and drops IPv6, the engine being IPv4-only). Nothing bypasses the TUN.
//   * PERSISTENT "block when the VPN *process* is down" — this is the OS Always-on / On-Demand
//     system setting. An app CANNOT silently enable it; `openVpnSettings` only deep-links the user
//     there, and the UI must say so. There is deliberately NO `block_without_vpn` argument (an
//     earlier one was hardcoded `true`, never read natively, and implied a control that didn't
//     exist — removed; see `extension.rs` and `DEVICE_VERIFY.md`).

import Foundation
import NetworkExtension
import Tauri
import UIKit
import WebKit

/// The attested start args handed over privately by `extension_connect` (Rust). Field names are camelCase to
/// match `extension::StartArgs` (`#[serde(rename_all = "camelCase")]`) and are forwarded VERBATIM
/// into the appex's `providerConfiguration`, where `PacketTunnelProvider.swift` reads the same keys.
/// There is no `allowUnattested` field on the real path — its absence means attestation is enforced.
struct NilStartTcbFloor: Decodable {
  let fmc: UInt8?
  let bootloader: UInt8
  let tee: UInt8
  let snp: UInt8
  let microcode: UInt8
}

class NilStartArgs: Decodable {
  let reservationId: String
  let nodeHost: String
  let nodePort: Int
  let serverName: String
  let measurementHex: String
  let tlsSpkiSha256Hex: String
  let transparencyLogKeyHex: String
  let teeName: String
  let minTcbSevsnp: NilStartTcbFloor?
  let grantHex: String
  let grantNonceHex: String
}

class NilVpnPlugin: Plugin {
  /// The packet-tunnel appex bundle id — a child of the container app id (`com.nilvpn.client`),
  /// matching `PacketTunnel-iOS-Info.plist`'s `CFBundleIdentifier`.
  private static let providerBundleId = "com.nilvpn.client.PacketTunnel"
  /// Settings-facing name of the VPN configuration this plugin manages.
  private static let vpnDescription = "NIL VPN"

  // MARK: - prepareVpn
  // Ensure the VPN configuration exists and OS consent is granted, WITHOUT starting the tunnel — so
  // a token is never redeemed before the user has approved the VPN. iOS has no separate
  // `VpnService.prepare()` (as Android does): the system "VPN Configurations" consent dialog is
  // presented by `saveToPreferences` the first time. We persist a minimal manager (triggering that
  // dialog if needed) and report whether it succeeded.
  @objc public func prepareVpn(_ invoke: Invoke) {
    loadOrCreateManager { manager, error in
      if let error = error {
        invoke.reject("VPN preflight failed: \(error.localizedDescription)")
        return
      }
      guard let manager = manager else {
        invoke.reject("VPN preflight failed: no manager")
        return
      }
      manager.localizedDescription = NilVpnPlugin.vpnDescription
      manager.isEnabled = true
      if manager.protocolConfiguration == nil {
        let proto = NETunnelProviderProtocol()
        proto.providerBundleIdentifier = NilVpnPlugin.providerBundleId
        proto.serverAddress = "NIL"  // display-only placeholder; startVpn sets the real endpoint
        manager.protocolConfiguration = proto
      }
      manager.saveToPreferences { saveError in
        // A save error is, in practice, the user declining the consent dialog — report it as
        // unauthorized (the frontend must not redeem a token) rather than a hard failure.
        invoke.resolve(["authorized": saveError == nil])
      }
    }
  }

  // MARK: - startVpn
  // Configure the manager with the attested args and start the tunnel. Resolves as soon as the start
  // is requested; the tunnel state is asynchronous, so the WebView polls `statusVpn` until "up"
  // (which only happens after the appex passes attestation and applies its network settings).
  @objc public func startVpn(_ invoke: Invoke) {
    let args: NilStartArgs
    do {
      args = try invoke.parseArgs(NilStartArgs.self)
    } catch {
      invoke.reject("invalid startVpn args: \(error.localizedDescription)")
      return
    }
    guard args.reservationId.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil else {
      invoke.reject("invalid reservation id")
      return
    }
    loadOrCreateManager { manager, error in
      if let error = error {
        invoke.reject("load VPN manager: \(error.localizedDescription)")
        return
      }
      guard let manager = manager else {
        invoke.reject("load VPN manager: none")
        return
      }
      let proto = NETunnelProviderProtocol()
      proto.providerBundleIdentifier = NilVpnPlugin.providerBundleId
      // `serverAddress` is display-only in Settings — never identity. Use the node host.
      proto.serverAddress = args.nodeHost
      // The keys the appex reads out of `providerConfiguration` (camelCase, matching
      // `PacketTunnelProvider.swift`). `nodePort` fits a u16 but rides as an Int (a JSON number).
      var providerConfiguration: [String: Any] = [
        "reservationId": args.reservationId,
        "nodeHost": args.nodeHost,
        "nodePort": args.nodePort,
        "serverName": args.serverName,
        "measurementHex": args.measurementHex,
        "tlsSpkiSha256Hex": args.tlsSpkiSha256Hex,
        "transparencyLogKeyHex": args.transparencyLogKeyHex,
        "teeName": args.teeName,
        "grantHex": args.grantHex,
        "grantNonceHex": args.grantNonceHex,
      ]
      if let floor = args.minTcbSevsnp {
        var encoded: [String: Any] = [
          "bootloader": Int(floor.bootloader),
          "tee": Int(floor.tee),
          "snp": Int(floor.snp),
          "microcode": Int(floor.microcode),
        ]
        if let fmc = floor.fmc { encoded["fmc"] = Int(fmc) }
        providerConfiguration["minTcbSevsnp"] = encoded
      }
      proto.providerConfiguration = providerConfiguration
      // In-connection fail-closed: route ALL traffic into the tunnel (protocol-level reinforcement of
      // the appex's default routes). The persistent "block when the process is down" guarantee is the
      // OS Always-on setting, which the app cannot set — see `openVpnSettings` (PD-8).
      proto.includeAllNetworks = true
      proto.excludeLocalNetworks = false
      manager.protocolConfiguration = proto
      manager.localizedDescription = NilVpnPlugin.vpnDescription
      manager.isEnabled = true
      // Save → reload (Apple requires a reload to obtain a valid session) → start.
      manager.saveToPreferences { saveError in
        if let saveError = saveError {
          invoke.reject(
            "save VPN preferences (grant the VPN permission, then connect again): \(saveError.localizedDescription)")
          return
        }
        manager.loadFromPreferences { loadError in
          if let loadError = loadError {
            invoke.reject("reload VPN preferences: \(loadError.localizedDescription)")
            return
          }
          do {
            try manager.connection.startVPNTunnel()
            invoke.resolve()
          } catch {
            invoke.reject("start tunnel: \(error.localizedDescription)")
          }
        }
      }
    }
  }

  // MARK: - stopVpn
  // Stop the tunnel. Resolves even if nothing is running (idempotent), mirroring Android.
  @objc public func stopVpn(_ invoke: Invoke) {
    loadOrCreateManager { manager, _ in
      manager?.connection.stopVPNTunnel()
      invoke.resolve()
    }
  }

  // MARK: - statusVpn
  // Map the REAL `NEVPNStatus` to the engine-state vocabulary the WebView expects
  // ("up" | "connecting" | "down"). `.connected` is reported only AFTER the appex calls its
  // `startTunnel` completion handler — which `PacketTunnelProvider.applySettingsAndRead()` does only
  // once attestation has passed — so "up" is a truthful "attested + connected", not optimistic.
  @objc public func statusVpn(_ invoke: Invoke) {
    loadOrCreateManager { manager, _ in
      let state: String
      switch manager?.connection.status {
      case .some(.connected):
        state = "up"
      case .some(.connecting), .some(.reasserting):
        state = "connecting"
      case .some(.disconnecting), .some(.disconnected), .some(.invalid), .none:
        state = "down"
      @unknown default:
        state = "down"
      }
      var response: [String: Any] = ["state": state]
      if let proto = manager?.protocolConfiguration as? NETunnelProviderProtocol,
         let reservationId = proto.providerConfiguration?["reservationId"] as? String,
         reservationId.range(of: "^[0-9a-f]{64}$", options: .regularExpression) != nil
      {
        response["reservationId"] = reservationId
      }
      invoke.resolve(response)
    }
  }

  // MARK: - openVpnSettings
  // Deep-link to OS settings so the user can enable the PERSISTENT kill-switch (Always-on /
  // Connect-On-Demand). iOS has NO public deep-link to Settings ▸ General ▸ VPN (Android has
  // `ACTION_VPN_SETTINGS`), and the private "App-Prefs:" scheme is App-Review-unsafe; the app's own
  // Settings page (`openSettingsURLString`, public + allowed) is the honest target, and the UI must
  // explain that the user enables Always-on there. The app cannot enable it silently (PD-8).
  @objc public func openVpnSettings(_ invoke: Invoke) {
    DispatchQueue.main.async {
      if let url = URL(string: UIApplication.openSettingsURLString) {
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
      }
      invoke.resolve()
    }
  }

  // MARK: - helpers
  /// Load only NIL's own manager or hand back a fresh one. Selecting `managers.first` could
  /// overwrite an unrelated packet-tunnel provider installed by another app/profile.
  private func loadOrCreateManager(_ completion: @escaping (NETunnelProviderManager?, Error?) -> Void) {
    NETunnelProviderManager.loadAllFromPreferences { managers, error in
      if let error = error {
        completion(nil, error)
        return
      }
      let existing = managers?.first { manager in
        guard let proto = manager.protocolConfiguration as? NETunnelProviderProtocol else {
          return false
        }
        return proto.providerBundleIdentifier == NilVpnPlugin.providerBundleId
      }
      completion(existing ?? NETunnelProviderManager(), nil)
    }
  }
}

/// Tauri's iOS registration entry point. The Rust side binds this symbol via
/// `tauri::ios_plugin_binding!(init_nil_vpn_plugin)` and hands it to `register_ios_plugin`
/// (see `client/src-tauri/src/lib.rs`). It returns the `Plugin` instance Tauri then manages.
@_cdecl("init_nil_vpn_plugin")
func initNilVpnPlugin() -> Plugin {
  return NilVpnPlugin()
}
